tutorial/ffnet_agent/CPUFFNetAgent.cpp

```diff
@@ -0,0 +1,431 @@
+    }
+}
+
+// Distribute incoming policy to children
+void CPUFFNetAgent::split_policy(const std::vector<double>& in_policy,
+                                std::vector<std::vector<double> >& out_policy)
+{
+    assert(in_policy.size() == M_NUM_POLICY);
+    for (auto &child_pol : out_policy) {
+        child_pol = in_policy;
+    }
+}
+
+// Indicate whether to send the policy down to children
+bool CPUFFNetAgent::do_send_policy(void) const
+{
+    return true;
+}
+
+void CPUFFNetAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                    std::vector<double>& out_sample)
+{
+
+}
+
+// Indicate whether to send samples up to the parent
+bool CPUFFNetAgent::do_send_sample(void) const
+{
+    return false;
+}
+
+void CPUFFNetAgent::adjust_platform(const std::vector<double>& in_policy)
+{
+    assert(in_policy.size() == M_NUM_POLICY);
+
+    m_do_write_batch = false;
+
+    // Per freq
+    std::vector<double> package_freq_request;
+    for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+        //Create an input tensor
+        // TODO
+        Vector xs(11);
+
+        xs[0] = (double) m_package_power.at(domain_idx).signal;
+        xs[1] = (double) m_package_freq_status.at(domain_idx).signal;
+        xs[2] = (double) m_package_temperature.at(domain_idx).signal;
+        xs[3] = (double) m_package_uncore_freq_status.at(domain_idx).signal;
+        xs[4] = (double) m_package_qm_ctr.at(domain_idx).signal;
+        xs[4] = (double) m_package_qm_rate.at(domain_idx).signal;
+
+        //Division of diffed values handled as part of the NN input
+        xs[5] = (double) m_package_inst_retired.at(domain_idx).sample /
+                         m_package_cycles_unhalted.at(domain_idx).sample;
+
+        xs[6] = (double) m_package_inst_retired.at(domain_idx).sample /
+                         m_package_energy.at(domain_idx).sample;
+
+        xs[7] = (double) m_package_acnt.at(domain_idx).sample /
+                         m_package_mcnt.at(domain_idx).sample;
+
+        xs[8] = (double) m_package_pcnt.at(domain_idx).sample /
+                         m_package_mcnt.at(domain_idx).sample;
+
+        xs[9] = (double) m_package_pcnt.at(domain_idx).sample /
+                         m_package_acnt.at(domain_idx).sample;
+
+        xs[10] = (double) in_policy[M_POLICY_CPU_PHI];
+
+        Vector output = m_package_neural_net.model(xs);
+
+        //Save recommendation and convert to SI units for later application
+        package_freq_request.push_back(output[0] * 1e9);
+//        std::cout << "CPUFFNet requested " << output[0] * 1e9 << std::endl;
+    }
+
+    // set frequency control per accelerator
+    for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+        //NAN --> Max Frequency
+        if(std::isnan(package_freq_request.at(domain_idx))) {
+            package_freq_request.at(domain_idx) = in_policy[M_POLICY_CPU_FREQ_MAX];
```

Comment:
If CPU_FREQ_MAX isn't used for anything other than the NAN case is there a reason to provide it in via policy vs relying on the CPU_FREQUENCY_MAX_AVAIL signal?